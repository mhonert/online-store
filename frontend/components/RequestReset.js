import React, { useState } from 'react';
import gql from 'graphql-tag';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';
import { Mutation } from 'react-apollo';
import { CURRENT_USER_QUERY } from './User';
import Router from 'next/router';

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
        message
    }
  }
`;

const RequestReset = () => {
  const [email, setEmail] = useState('');

  const handleSubmit = async (e, requestReset) => {
    e.preventDefault();

    const response = await requestReset();
    setEmail('');
  };

  return (
    <Mutation
      mutation={REQUEST_RESET_MUTATION}
      variables={{ email }}
    >
      {(requestReset, { error, loading, called }) => {
        return (
          <Form method="post" onSubmit={e => handleSubmit(e, requestReset)}>
            <fieldset disabled={loading} aria-busy={loading}>
              <h2>Forgot password?</h2>
              <DisplayError error={error} />
              {!error && !loading && called && <p>Success! Check your email for a reset link.</p>}
              <label>
                Email
                <input
                  type="email"
                  name="email"
                  placeholder="Email"
                  value={email}
                  required
                  onChange={e => setEmail(e.target.value)}
                />
              </label>
            </fieldset>

            <button type="submit">Reset Password!</button>
          </Form>
        );
      }}
    </Mutation>
  );
};

export default RequestReset;
