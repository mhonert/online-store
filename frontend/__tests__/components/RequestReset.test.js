import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';

import { MockedProvider } from 'react-apollo/test-utils';
import React from 'react';
import RequestReset, {
  REQUEST_RESET_MUTATION
} from '../../components/RequestReset';

const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email: 'test@localhost' }
    },
    result: {
      data: {
        requestReset: { message: 'success', __typename: 'Message' }
      }
    }
  }
];

describe('<RequestReset />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );

    expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
  });

  it('calls the mutation', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );

    // simulate entered email address
    wrapper.find('input').simulate('change', {
      target: { name: 'email', value: 'test@localhost' }
    });

    await wait();
    wrapper.update();

    wrapper.find('form').simulate('submit');
    await wait();
    wrapper.update();

    expect(wrapper.find('p').text()).toContain(
      'Success! Check your email for a reset link'
    );
  });
});
