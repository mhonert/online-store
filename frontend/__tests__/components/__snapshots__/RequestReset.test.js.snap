// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-3ek8g3-0 kxkTjx"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Forgot password?
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label>
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        value=""
      />
    </label>
  </fieldset>
  <button
    type="submit"
  >
    Reset Password!
  </button>
</form>
`;
